File: frontend/lib/models/transaction.dart
class Transaction {
  final String id;
  final String bankName;
  final String accountType;
  final String transactionDate;
  final String description;
  final double amount;
  final String type;
  String category;
  String classification;
  final String? merchantCategory;
  final bool isEmi;
  final bool isRecurring;

  Transaction({
    required this.id,
    required this.bankName,
    required this.accountType,
    required this.transactionDate,
    required this.description,
    required this.amount,
    required this.type,
    required this.category,
    required this.classification,
    this.merchantCategory,
    this.isEmi = false,
    this.isRecurring = false,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      bankName: json['bank_name'] as String? ?? '',
      accountType: json['account_type'] as String? ?? '',
      transactionDate: json['transaction_date'] as String? ?? '',
      description: json['description'] as String? ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      type: json['type'] as String? ?? 'debit',
      category: json['category'] as String? ?? 'Uncategorized',
      classification: json['classification'] as String? ?? 'uncategorized',
      merchantCategory: json['merchant_category'] as String?,
      isEmi: json['is_emi'] as bool? ?? false,
      isRecurring: json['is_recurring'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bank_name': bankName,
      'account_type': accountType,
      'transaction_date': transactionDate,
      'description': description,
      'amount': amount,
      'type': type,
      'category': category,
      'classification': classification,
      'merchant_category': merchantCategory,
      'is_emi': isEmi,
      'is_recurring': isRecurring,
    };
  }

  bool get isDebit => type == 'debit';
  bool get isCredit => type == 'credit';
  String get formattedAmount {
    return '₹${amount.toStringAsFixed(0)}';
  }
}

File: backend/app/schemas/transaction.py
from pydantic import BaseModel, Field, field_validator, ConfigDict
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

VALID_CLASSIFICATIONS = {"mandatory", "discretionary", "income", "uncategorized"}


class ManualTransactionCreate(BaseModel):
    transaction_date: date
    description: str = Field(..., min_length=1, max_length=1000)
    amount: Decimal = Field(..., gt=0, decimal_places=2)
    bank_name: str = Field(default="Manual", max_length=100)
    account_type: str = Field(default="manual", max_length=50)
    type: str = Field(default="debit", pattern=r"^(debit|credit)$")
    merchant_category: str | None = Field(default=None, max_length=200)
    is_emi: bool = False
    is_recurring: bool = False


class TransactionUpdate(BaseModel):
    category: str | None = Field(default=None, max_length=100)
    classification: str | None = Field(default=None, max_length=50)
    tags: str | None = Field(default=None, max_length=500)
    type: str | None = Field(default=None, pattern=r"^(debit|credit)?$")
    description: str | None = Field(default=None, max_length=1000)

    @field_validator("classification")
    @classmethod
    def validate_classification(cls, v: str | None) -> str | None:
        if v is not None and v not in VALID_CLASSIFICATIONS:
            raise ValueError(f"classification must be one of {VALID_CLASSIFICATIONS}")
        return v


class TransactionCreate(BaseModel):
    bank_name: str
    account_type: str = "credit_card"
    source: str = "pdf_upload"
    transaction_date: date
    description: str
    amount: Decimal
    type: str = "debit"
    category: str = "Uncategorized"
    classification: str = "uncategorized"
    merchant_category: Optional[str] = None
    is_emi: bool = False
    is_recurring: bool = False
    tags: Optional[str] = None
    statement_file: Optional[str] = None


class TransactionResponse(TransactionCreate):
    id: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class TransactionFilter(BaseModel):
    bank_name: Optional[str] = None
    category: Optional[str] = None
    classification: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    min_amount: Optional[Decimal] = None
    max_amount: Optional[Decimal] = None
    search: Optional[str] = None
    type: Optional[str] = None


class CategorySummary(BaseModel):
    category: str
    classification: str
    total: Decimal
    count: int
    percentage: float


class MonthlyTrend(BaseModel):
    month: str
    mandatory: Decimal
    discretionary: Decimal
    total: Decimal


class Insight(BaseModel):
    title: str
    description: str
    savings_potential: Optional[Decimal] = None
    category: Optional[str] = None


class AnalysisResponse(BaseModel):
    total_spend: Decimal
    mandatory_spend: Decimal
    discretionary_spend: Decimal
    mandatory_percentage: float
    discretionary_percentage: float
    monthly_trends: list[MonthlyTrend]
    category_breakdown: list[CategorySummary]
    insights: list[Insight]
